NDTensors/src/lib/UnallocatedArrays/src/abstractunallocatedarray.jl

```diff
@@ -0,0 +1,65 @@
+using FillArrays: FillArrays, getindex_value
+using NDTensors.SetParameters: set_parameters
+using Adapt: adapt
+
+const UnallocatedArray{ElT,N,AxesT,AllocT} = Union{
+  UnallocatedFill{ElT,N,AxesT,AllocT},UnallocatedZeros{ElT,N,AxesT,AllocT}
+}
+
+@inline Base.axes(A::UnallocatedArray) = axes(parent(A))
+Base.size(A::UnallocatedArray) = size(parent(A))
+function FillArrays.getindex_value(A::UnallocatedArray)
+  return getindex_value(parent(A))
+end
+
+function Base.complex(A::UnallocatedArray)
+  return complex(eltype(A)).(A)
+end
+
+function Base.transpose(a::UnallocatedArray)
+  return set_alloctype(transpose(parent(a)), alloctype(a))
+end
+
+function Base.adjoint(a::UnallocatedArray)
+  return set_alloctype(adjoint(parent(a)), alloctype(a))
+end
+
+function set_alloctype(T::Type{<:UnallocatedArray}, alloc::Type{<:AbstractArray})
+  return set_parameters(T, Position{4}(), alloc)
+end
+
+## This overloads the definition defined in `FillArrays.jl`
+for STYPE in (:AbstractArray, :AbstractFill)
+  @eval begin
+    @inline $STYPE{T}(F::UnallocatedArray{T}) where {T} = F
+    @inline $STYPE{T,N}(F::UnallocatedArray{T,N}) where {T,N} = F
+  end
+end
+
+function allocate(f::UnallocatedArray)
+  a = similar(f)
+  fill!(a, getindex_value(f))
+  return a
+end
+
+function allocate(arraytype::Type{<:AbstractArray}, elt::Type, axes::Tuple)
+  ## TODO rewrite this using set_eltype and set_ndims functions
+  ## currently these functions are defined in `NDTensors`
+  ## In the future they should be defined in `SetParameters`
+  ArrayT = set_parameters(arraytype, Position{1}(), elt)
+  ArrayT = set_parameters(ArrayT, Position{2}(), length(axes))
+  return similar(ArrayT, axes)
+end
+
+function Base.similar(f::UnallocatedArray)
+  return allocate(alloctype(f), eltype(f), axes(f))
+end
```

Comment:
If you look at https://docs.julialang.org/en/v1/manual/interfaces/#man-interface-array, you can see that `similar(a)` by default calls `similar(a, eltype(a), size(a))`, so it's best to define that version.

So:
```julia
function Base.similar(f::UnallocatedArray, elt::Type, axes)
  return allocate(alloctype(f), elt, axes)
end
```
and Julia should define `Base.similar(f::UnallocatedArray)` automatically.

Reply:
Okay this makes sense. However if I define the function as `function Base.similar(f::UnallocatedArray, elt::Type, axes)`, the code can fail because of ambiguity
```
: MethodError: similar(::UnallocatedZeros{Float32, 2, Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}}, Type{Matrix{Float32}}}, ::Type{Float32}, ::Tuple{Int64, Int64}) is ambiguous.
```
the compiler suggests defining the function like 
```
function Base.similar(f::UnallocatedArray, elt::Type, axes::Tuple{Int64, Vararg{Int64}})
```
defining it like this no longer throws an error and the code runs as expected